### PR TITLE
Add CBMC driver program readme

### PR DESCRIPTION
**Pull request by smowton**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: bb081fd7562839d3fcf961c3d6a1472060a07082).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91240341

File: src/cbmc/README.md [hunk 1]
```diff
@@ -11,3 +11,106 @@ others) to create a goto-program, `goto-symex` to unwind the loops the
...
+
+The class `bmct` is responsible for solving the equation, generated by
+goto-symex, which represents the assertions or goals it is being asked to show
+can be hit or satisfied (or prove that can't be done).
+
+For example, given an input program
+`f(int x, int y) { assert(x + y % 3 == 1); assert(x % y == 0); }` then goto-
+symex would produce a formula something like
+`F1 + F2 % 3 != 1 OR (F1 + F2 % 3 == 1 AND F1 % F2 != 0), where F-variables
+are free. The precise expression of that formula depends on cbmc's command-line
```
Comment (peterschrammel) [inline src/cbmc/README.md]: CBMC

File: src/cbmc/README.md [hunk 2]
```diff
@@ -11,3 +11,106 @@ others) to create a goto-program, `goto-symex` to unwind the loops the
...
+
+## Path-symex
+
+CBMC supports running goto-symex in "path explorer" mode, which alters the
+simple GOTO model, goto-symex, equation solving sequence described above by
+incrementally running goto-symex to produce a formula describing *part* of a
+program, trying to solve that partial formula to satisfy some properties (or
+achieve some coverage, as appropriate), before re-entering symex to examine
+other properties. For example, given a simple program
+`if(x) assert(y) else assert(z)`, standard CBMC would generate a single formula
```
Comment (peterschrammel) [inline src/cbmc/README.md]: if(x) assert(y); else assert(z);

**Review by peterschrammel — APPROVED** [review]

File: src/cbmc/README.md [hunk 3]
```diff
@@ -11,3 +11,106 @@ others) to create a goto-program, `goto-symex` to unwind the loops the
...
+
+1. Parse command-line options; convert passed source files into a GOTO model
+2. Run `goto-symex` to turn the GOTO model into a SAT or SMT equation
+3. Solve the equation and report results
+
+The GOTO model is generated using `initialize_goto_model`; see documentaton
+in the `goto-programs/` directory. Goto-symex is similarly documented in its
+own directory.
+
+## Equation solving
```
Comment (tautschnig) [inline src/cbmc/README.md:30]: Shouldn't references to the documentation @johanneskloos is working on also be provided here? It's good that the documentation here focuses on the reports generated and the options, but the reader may be curious about the details of doing bounded model checking.

Comment (smowton) [reply src/cbmc/README.md:30]: Unsure how to do that exactly, but sure. @johanneskloos any hints?

Comment (johanneskloos) [reply src/cbmc/README.md:30]: The general approach behind most of the steps is documented in background concepts (background-concepts should work for a link); in particular, translation to an IR (not quite GOTO, but in the same general spirit) is in IR_section, while steps 2 and 3 are described in BMC_section (referencing the SSA section). Does this help?

**Comment by smowton** [comment]
@tautschnig added some links

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fa595d8dbf74d3d1a59c2f4d68bbc70118a4f665).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91397409

**Pull request merged**